### PR TITLE
pre-commit check for xfail in unit tests

### LINKED ISSUE #28768: Add pre-commit to prevent using xfail tests
### Body

Related https://github.com/apache/airflow/issues/23539

xfail means that you expect a test to fail for some reason.
If we expect a test to fail it should simply be a regular test that make sure the relevant logic fails and in that case we finish the test with success.|
There should be no cases where we see tests ends in failure yet we consider it good - this is very confusing.

This task is about writing pre-commit that will make sure we don't add any xfail test to the project.

### Committer

- [X] I acknowledge that I am a maintainer/committer of the Apache Airflow project.

### PR DESCRIPTION
**Pull request by frankcash**
will resolve #28768 

Adds regexp check for `@pytest.mark.xfail` in files `^tests/.*\.py$`

Screen Shot 2023-01-06 at 3 23 24 PM

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Comment by eladkal** [comment]
We still have tests that needs to be excluded https://github.com/apache/airflow/issues/23539 as we didnt finish removing xfail everywhere

**Review by eladkal — APPROVED** [review]

**Comment by frankcash** [comment]
@eladkal we could also wait to merge this until https://github.com/apache/airflow/pull/28724 is merged

**Comment by Taragolis** [comment]
We also have one xfail marked test which not listed in https://github.com/apache/airflow/issues/23539

https://github.com/apache/airflow/blob/3eee33ac8cb74cfbb08bce9090e9c601cf98da44/tests/always/test_project_structure.py#L377-L379

One of suggestion is remove this kind of tests, see: https://github.com/apache/airflow/pull/28459#discussion_r1054681194

**Comment by eladkal** [comment]
@frankcash I suggest to add exclusion for the 1-2 remaining tests
cleaning it after tests are refactored is simple

**Comment by frankcash** [comment]
@eladkal does this look like adding those files specifically to the exclude file list? `        exclude: ^airflow/|^docs/`

**Comment by frankcash** [comment]
@eladkal ✅

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -577,6 +577,14 @@ repos:
...
         exclude: ^airflow/_vendor/
         pass_filenames: true
+      - id: check-xfail-usage-in-test-code
+        language: pygrep
+        name: Don't use xfail in tests
+        description: xfail is not recommended as a test meant to fail is generally not meaningful
+        entry: "@pytest.mark.xfail"
+        pass_filenames: true
+        files: ^tests/.*\.py$
+        exclude: ^airflow/|^docs/|^tests/always/test_project_structure.py|^tests/jobs/test_backfill_job.py
```
Comment (jedcunningham) [inline .pre-commit-config.yaml]: Why do we need `^airflow` and `^docs` here, if we are only matching `^tests/.*\.py$`?

Comment (frankcash) [reply .pre-commit-config.yaml]: good call out, that was just copy pasta from before, will remove it now :)

**Comment by ashb** [comment]
This is a -0.99 from me. XFail exists for a reason, we can discourage it's use without needing to remove it.

File: .pre-commit-config.yaml (hunk 1 above)
Comment (frankcash) [reply .pre-commit-config.yaml]: ✅

**Comment by potiuk** [comment]
> This is a -0.99 from me. XFail exists for a reason, we can discourage it's use without needing to remove it.

Yep. I think there are reasons why you might want to have xfail. Some of the  other fixes xfails in #23539  were simply mis-used - their reason was  "the tests were flaky" but using xfail for the reason of "we know it should fail and we can fix it later" is quite ok.

**Comment by potiuk** [comment]
BTW. it does not mean that the encouragement of getting rid of those xfals is bad idea (if done correctly).

**Comment by ashb** [comment]
Oh yeah I'm cool with "don't use XFail for flaky tests" etc.

**Pull request closed without merging**